src/runtime/lib/telemetry/index.ts

```diff
@@ -85,62 +103,122 @@ export class TelemetryClient {
         this.enabled = this.isEnabled();
+        const osType = os.type();
+        const osArch = os.arch();
+        const osRelease = os.release();
+        let gitVersion = '';
+        try {
+            const gitVersionUint8Array: Uint8Array = child_process.execSync('git --version');
+            gitVersion = new TextDecoder(DEFAULT_ENCODING).decode(gitVersionUint8Array);
+        } catch (error) {
+            loggableAskError('TelemetryClient failed to fetch git version', error);
+        }
+        
         this.data = {
-            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
             version: extensions.getExtension(EXTENSION_ID)?.packageJSON.version,
             machineId: env.machineId,
             timeStarted: new Date(),
             newUser: false,
             timeUploaded: null,
             clientId: EXTENSION_FULL_NAME,
+            operatingSystem: `${osType} ${osArch} ${osRelease}`,
+            dependencyVersion: version,
+            gitVersion: gitVersion,
             actions: []
         };
     }
 
+    /**
+     * The static method that controls the access to the TelemetryClient instance
+     * @param options 
+     * @returns 
+     */
+    public static getInstance(options?: TelemetryClientOptions) {
+        if (TelemetryClient.instance == undefined) {
+            TelemetryClient.instance = new TelemetryClient(options);
+        }
+        return TelemetryClient.instance;
+    }
+
     /**
      * Starts action
      * @param {string} name - The action name
      * @param {string} type - The action type
      * @return {MetricAction}
      */
-    public startAction(name: string, type: string): MetricAction{
-        const action = new MetricAction(name, type);
-        this.data.actions.push(action);
-        return action;
+    public startAction(name: string, type: ActionType): MetricAction {
+        return new MetricAction(name, type);
     }
 
     /**
-     * Sends data to the metric server
-     * @param {Error} [error] - Error object indicating an error.
-     * @returns {Promise<{success: boolean}>}
+     * Store an action and send the data
+     * @param action - The MetricAction item
+     * @param error - The error object
      */
-    public sendData(error?: Error): Promise<TelemetryUploadResult> {
+    public async store(action: MetricAction , error?:Error): Promise<void> {
+        action.end(error);
         if (!this.enabled) {
-            this.data.actions = [];
-            Logger.debug('Telemetry is disabled. Not uploading any data.');
+            await this.resetStoredStates();
+            loggableAskError('Telemetry is disabled. Not uploading any data.');
+            return;
+        }
+        let dataMap = ext.context.globalState.get(TELEMETRY_UPDATE_TIME_DATA_DIC) as {};
+        if (dataMap == undefined) {
+            const now = new Date().getTime();
+            const dic = {[now]: []}
```

Comment:
nit: Very generic name. Maybe put something specific. Or in fact, we do not even need to create this `const` and do direct assignment